src/test/app/AMM_test.cpp

```diff
@@ -0,0 +1,8180 @@
+                XRP(10000), USD(10201), IOUAmount{10100000, 0}));
+        });
+
+        // Single deposit: 100000 tokens worth of XRP
+        testAMM([&](AMM& ammAlice, Env&) {
+            ammAlice.deposit(carol, 100000, XRP(205));
+            BEAST_EXPECT(ammAlice.expectBalances(
+                XRP(10201), USD(10000), IOUAmount{10100000, 0}));
+        });
+
+        // Single deposit with EP not exceeding specified:
+        // 100USD with EP not to exceed 0.1 (AssetIn/TokensOut)
+        testAMM([&](AMM& ammAlice, Env&) {
+            ammAlice.deposit(
+                carol, USD(1000), std::nullopt, STAmount{USD, 1, -1});
+            BEAST_EXPECT(ammAlice.expectBalances(
+                XRP(10000), USD(11000), IOUAmount{1048808848170152, -8}));
+        });
+
+        // Single deposit with EP not exceeding specified:
+        // 100USD with EP not to exceed 0.002004 (AssetIn/TokensOut)
+        testAMM([&](AMM& ammAlice, Env&) {
+            ammAlice.deposit(
+                carol, USD(100), std::nullopt, STAmount{USD, 2004, -6});
+            BEAST_EXPECT(ammAlice.expectBalances(
+                XRP(10000),
+                STAmount{USD, 1008016, -2},
+                IOUAmount{10040000, 0}));
+        });
+
+        // Single deposit with EP not exceeding specified:
+        // 0USD with EP not to exceed 0.002004 (AssetIn/TokensOut)
+        testAMM([&](AMM& ammAlice, Env&) {
+            ammAlice.deposit(
+                carol, USD(0), std::nullopt, STAmount{USD, 2004, -6});
+            BEAST_EXPECT(ammAlice.expectBalances(
+                XRP(10000),
+                STAmount{USD, 1008016, -2},
+                IOUAmount{10040000, 0}));
+        });
+
+        // IOU to IOU + transfer fee
+        {
+            Env env{*this};
+            fund(env, gw, {alice}, {USD(25000), BTC(0.625)}, Fund::All);
+            env(rate(gw, 1.25));
+            env.close();
+            AMM ammAlice(env, alice, USD(20000), BTC(0.5));
+            BEAST_EXPECT(ammAlice.expectBalances(
+                USD(20000), BTC(0.5), IOUAmount{100, 0}));
+            BEAST_EXPECT(expectLine(env, alice, USD(0)));
+            BEAST_EXPECT(expectLine(env, alice, BTC(0)));
+            fund(env, gw, {carol}, {USD(2500), BTC(0.0625)}, Fund::Acct);
+            ammAlice.deposit(carol, 10);
+            BEAST_EXPECT(ammAlice.expectBalances(
+                USD(22000), BTC(0.55), IOUAmount{110, 0}));
+            // 2,500 - 2,000(AMM) - 0.25*2,000=500(fee)=0
+            BEAST_EXPECT(expectLine(env, carol, USD(0)));
+            // 0.0625 - 0.05(AMM) - 0.25*0.05=0.0125(fee)=0
+            BEAST_EXPECT(expectLine(env, carol, BTC(0)));
+        }
+
+        // Tiny deposits
+        testAMM([&](AMM& ammAlice, Env&) {
+            ammAlice.deposit(carol, IOUAmount{1, -3});
+            BEAST_EXPECT(ammAlice.expectBalances(
+                XRPAmount{10000000001},
+                STAmount{USD, UINT64_C(10000000001), -6},
```

Comment:
Rather than `UINT64_C`, the convention in the rest of the code is the `ull` suffix. So `10000000001ull`. I notice this is used several times in this file. Please change them all, and I won't make you read a duplicate comment on every one.

Reply:
Fascinating. Visual Studio treats `std::uint64_t` and `unsigned long long` as the same type.
```
#define UINT64_C(x)  (x ## ULL)
```
and
```
typedef unsigned long long uint64_t;
```

When I compile that snippet in VS (don't mind the line numbers):
```
Main.cpp(818,1): error C2084: function 'void foo(uint64_t,int)' already has a body
Main.cpp(794): message : see previous definition of 'foo'
Main.cpp(846,29): error C2264: 'foo': error in function definition or declaration; function not called
```
So in this code, if I comment out `foo(unsigned long long m, int e)`, then the output is
```
same uint64_t
same unsigned long long and std::uint64_t
uint64_t
```
Which is quite different than what `gcc` outputs:
```
same unsigned long long
unsigned long long
```

So clearly without a `std::uint64_t` suffix, your code is more correct.

What do you think about adding `STAmount` ctor overloads for `unsigned long long` if the two types _aren't_ the same?

Reply:
I don't want to touch `STAmount` unless I really must. I would have considered it, had it been required in the main code. I think using `UINT64_C` in the unit-test solves the problem and is fine.